### ISSUE TITLE
R2s from .predict() and .test() do not match

### ISSUE THREAD

**Issue by giusmatera** [Lightning-AI/torchmetrics#2874]
Hi, I found what it seems to be a bug:

I have
```py
from torchmetrics import MeanSquaredError, R2Score

	def test_step(self, batch, batch_idx):
	
	    batch = batch.to(torch.float32)
	
	    pooled_output = batch[
	        :, :-1
	    ].float()  # All columns except the last
	    targets = batch[:, -1].float()
	
	    predictions = self(pooled_output).squeeze(-1)
	
	    score_r2 = self.test_r2(predictions, targets)  # Compute R2
	    self.log("test_r2", score_r2, on_epoch=True)
	
	    return score_r2
```
but the test r2 score this step gives differs from what I retrieve directly from predictions and actual values.

To clarify, the r2 score in `results = trainer.test(regressor_model, test_dataloader)` is different from what I get from
```py
predictions = trainer.predict(regressor_model, test_dataloader)
torchmetrics.R2Score()(predictions, error)
```
Also, torchmetrics and sklearn r2 functions match in the latter task. How come? Thanks a lot.

Info:
- torchmetrics' version is 1.4.2

**Comment by SkafteNicki** [comment]
Hi @giusmatera,
Sorry for the late reply from my side, but there is a good reason for the difference in value. When you write:
```python
score_r2 = self.test_r2(predictions, targets)  # Compute R2
self.log("test_r2", score_r2, on_epoch=True)
```
you are first calculating the r2 score at batch level and the you are logging that value using `self.log`. When `self.log` receives a value (in this case a tensor) it does not know where this value came from, therefore the only way it can aggregate it over all your batches are using simple aggregation functions. `self.log` uses `torch.mean` by default
https://lightning.ai/docs/pytorch/stable/common/lightning_module.html#log
If you know anything about the `r2` score you will most likely know that you cannot calculate the overall r2 score as a average of the r2 score on separate batches. To get the correct value you need to aggregate the sum of residual squares and sum of total squares over all batches and then to the final 1-ss_res / ss_tot. To get the correct result you should instead just log the *metric object* and not the returned value
```python
self.test_r2(predictions, targets)  # updates the metric with batch data
self.log("test_r2", self.test_r2, on_epoch=True)
```
lightning will then log the correct value. You can read more here about logging torchmetrics in lightning
https://lightning.ai/docs/torchmetrics/stable/pages/lightning.html

Finally, the reason why `torchmetrics.R2Score()(predictions, error)` works is because you are feeding all data in as one batch and therefore the aggregation cannot go wrong.

Closing issue.
